enterprise/tailnet/pgcoord.go

```diff
@@ -106,30 +109,105 @@ func NewPGCoord(ctx context.Context, logger slog.Logger, ps pubsub.Pubsub, store
 	id := uuid.New()
 	logger = logger.Named("pgcoord").With(slog.F("coordinator_id", id))
 	bCh := make(chan binding)
-	cCh := make(chan *connIO)
+	cCh := make(chan agpl.Queue)
+	sCh := make(chan subscribe)
 	// signals when first heartbeat has been sent, so it's safe to start binding.
 	fHB := make(chan struct{})
 
 	c := &pgCoord{
-		ctx:            ctx,
-		cancel:         cancel,
-		logger:         logger,
-		pubsub:         ps,
-		store:          store,
-		binder:         newBinder(ctx, logger, id, store, bCh, fHB),
-		bindings:       bCh,
-		newConnections: cCh,
-		id:             id,
-		querier:        newQuerier(ctx, logger, ps, store, id, cCh, numQuerierWorkers, fHB),
-		closed:         make(chan struct{}),
+		ctx:              ctx,
+		cancel:           cancel,
+		logger:           logger,
+		pubsub:           ps,
+		store:            store,
+		binder:           newBinder(ctx, logger, id, store, bCh, fHB),
+		bindings:         bCh,
+		newConnections:   cCh,
+		subscriber:       newSubscriber(ctx, logger, id, store, sCh, fHB),
+		newSubscriptions: sCh,
+		id:               id,
+		querier:          newQuerier(ctx, logger, id, ps, store, id, cCh, numQuerierWorkers, fHB),
+		closed:           make(chan struct{}),
 	}
 	logger.Info(ctx, "starting coordinator")
 	return c, nil
 }
 
 func (c *pgCoord) ServeMultiAgent(id uuid.UUID) agpl.MultiAgentConn {
-	_, _ = c, id
-	panic("not implemented") // TODO: Implement
+	ma := (&agpl.MultiAgent{
+		ID:                id,
+		AgentIsLegacyFunc: func(agentID uuid.UUID) bool { return true },
+		OnSubscribe: func(enq agpl.Queue, agent uuid.UUID) (*agpl.Node, error) {
+			err := c.addSubscription(enq, agent)
+			return c.Node(agent), err
+		},
+		OnUnsubscribe: c.removeSubscription,
+		OnNodeUpdate: func(id uuid.UUID, node *agpl.Node) error {
+			return sendCtx(c.ctx, c.bindings, binding{
+				bKey: bKey{id, agpl.QueueKindClient},
+				node: node,
+			})
+		},
+		OnRemove: func(enq agpl.Queue) {
+			b := binding{
+				bKey: bKey{
+					id:   enq.UniqueID(),
+					kind: enq.Kind(),
+				},
+			}
+			if err := sendCtx(c.ctx, c.bindings, b); err != nil {
+				c.logger.Debug(c.ctx, "parent context expired while withdrawing binding", slog.Error(err))
+			}
+			if err := sendCtx(c.ctx, c.newSubscriptions, subscribe{
+				sKey:   sKey{clientID: id},
+				active: false,
+			}); err != nil {
+				c.logger.Debug(c.ctx, "parent context expired while withdrawing subscriptions", slog.Error(err))
+			}
+			c.querier.cleanupConn(enq)
```

Comment:
the querier waits for the context to be done and calls `cleanupConn()` -- so you end up cleaning up the connection twice.